site/src/contexts/ProxyContext.tsx

```diff
@@ -0,0 +1,211 @@
+import { useDashboard } from "components/Dashboard/DashboardProvider"
+import {
+  createContext,
+  FC,
+  PropsWithChildren,
+  useContext,
+  useState,
+} from "react"
+
+interface ProxyContextValue {
+  proxy: PreferredProxy
+  isLoading: boolean
+  error?: Error | unknown
+  setProxy: (regions: Region[], selectedRegion: Region | undefined) => void
+}
+
+interface PreferredProxy {
+  // SelectedRegion is the region the user has selected.
+  // Do not use the fields 'path_app_url' or 'wildcard_hostname' from this
+  // object. Use the preferred fields.
+  selectedRegion: Region | undefined
+  // PreferredPathAppURL is the URL of the proxy or it is the empty string
+  // to indicate using relative paths. To add a path to this:
+  //  PreferredPathAppURL + "/path/to/app"
+  preferredPathAppURL: string
+  // PreferredWildcardHostname is a hostname that includes a wildcard.
+  preferredWildcardHostname: string
+}
+
+export const ProxyContext = createContext<ProxyContextValue | undefined>(
+  undefined,
+)
+
+/**
+ * ProxyProvider interacts with local storage to indicate the preferred workspace proxy.
+ */
+export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
+  // Try to load the preferred proxy from local storage.
+  let savedProxy = loadPreferredProxy()
+  if (!savedProxy) {
+    savedProxy = getURLs([])
+  }
+
+  // The initial state is no regions and no selected region.
+  const [proxy, setProxy] = useState<PreferredProxy>(savedProxy)
+  const setAndSaveProxy = (
+    regions: Region[],
+    selectedRegion: Region | undefined,
+  ) => {
+    const preferred = getURLs(regions, selectedRegion)
+    // Save to local storage to persist the user's preference across reloads
+    // and other tabs.
+    savePreferredProxy(preferred)
+    // Set the state for the current context.
+    setProxy(preferred)
+  }
+
+  const queryKey = ["get-regions"]
+  const { error: regionsError, isLoading: regionsLoading } = useQuery({
+    queryKey,
+    queryFn: getWorkspaceProxies,
+    // This onSuccess ensures the local storage is synchronized with the
+    // regions returned by coderd. If the selected region is not in the list,
+    // then the user selection is removed.
+    onSuccess: (data) => {
+      setAndSaveProxy(data.regions, proxy.selectedRegion)
+    },
+  })
+
+  // ******************************* //
+  // ** This code can be removed  **
+  // ** when the experimental is  **
+  // **       dropped             ** //
+  const dashboard = useDashboard()
+  const appHostQueryKey = ["get-application-host"]
+  const {
+    data: applicationHostResult,
+    error: appHostError,
+    isLoading: appHostLoading,
+  } = useQuery({
+    queryKey: appHostQueryKey,
+    queryFn: getApplicationsHost,
+  })
+
+  // If the experiment is disabled, then make the setState do a noop.
+  // This preserves an empty state, which is the default behavior.
+  if (!dashboard?.experiments.includes("moons")) {
+    const value = getURLs([])
```

Comment:
Instead of defining a completely different `<ProxyContent.Provider />` here, we can use [our `createContext` hook](https://github.com/coder/coder/pull/7283/files#diff-1beec2c2724792e5a29bc182e07983cd672b52ea138eca517ddb65c5e30bb500R34) to define initial values, e.g:
````
export const ProxyContext = createContext<ProxyContextValue>({
  proxy: {...your default}
  isLoading: false
  setProxy: () => {}
})
````
Then we can bail out with an early return if the experiment isn't turned on. We want to avoid returning alternative `<ProxyContent.Provider />` elements.

Reply:
I'm not sure how to do this :thinking:. I need to set all the values except for the `SetProxy` function

Reply:
Fixed in chat with Kira